sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/RequestOptions.java

```diff
@@ -333,39 +254,36 @@ public InstrumentationContext getInstrumentationContext() {
      * @throws IllegalStateException if this instance is obtained by calling {@link RequestOptions#none()}.
      */
     public RequestOptions setInstrumentationContext(InstrumentationContext instrumentationContext) {
-        checkLocked("Cannot set instrumentation context.");
         this.instrumentationContext = instrumentationContext;
-
         return this;
     }
 
     /**
-     * Gets the {@link ProgressReporter} used to track progress of I/O operations of the request.
-     *
-     * @return The {@link ProgressReporter} used to track progress of I/O operations of the request.
-     */
-    public ProgressReporter getProgressReporter() {
-        return progressReporter;
-    }
-
-    /**
-     * Sets the {@link ProgressReporter} used to track progress of I/O operations of the request.
+     * Adds a key-value pair to the {@link RequestOptions} that can be used to pass additional data to the
+     * components of HTTP pipeline.
      *
-     * @param progressReporter The {@link ProgressReporter} used to track progress of I/O operations of the request.
+     * @param key The key to be added.
+     * @param value The value to be added.
      * @return The updated {@link RequestOptions} object.
      * @throws IllegalStateException if this instance is obtained by calling {@link RequestOptions#none()}.
+     * @throws NullPointerException If {@code key} is null.
      */
-    public RequestOptions setProgressReporter(ProgressReporter progressReporter) {
-        checkLocked("Cannot set progress reporter.");
-        this.progressReporter = progressReporter;
-
+    public RequestOptions putData(String key, Object value) {
```

Comment:
`RequestOptions.putData()` can be misinterpreted as the place where you can set request data (like request body).  

Other name suggestions: 
`putPipelineData` to indicate that this is mostly for client-side/pipeline only
`putRequestContextData` to indicate this data is request context and not the request itself 
`putRequestHandlingData` or `putRequestProcessingData` to indicate it's for handling the request (not sent to service).

Reply:
changed to `get|putMetadata`, lmk if it sounds reasonable